hibernate-core/src/main/java/org/hibernate/internal/util/QuotingHelper.java

```diff
@@ -70,13 +70,22 @@ public static String unquoteIdentifier(String text) {
 	}
 
 	public static String unquoteStringLiteral(String text) {
+		return unquoteString( text, false, 0 );
+	}
+
+	public static String unquoteJavaStringLiteral(String text) {
+		final char firstChar = text.charAt( 0 );
+		return unquoteString( text, true, firstChar == 'j' || firstChar == 'J' ? 1 : 0 );
+	}
+
+	public static String unquoteString(String text, boolean unescape, int start) {
 		assert text.length() > 1;
 		final int end = text.length() - 1;
-		final char delimiter = text.charAt( 0 );
+		final char delimiter = text.charAt( start );
 		assert delimiter == text.charAt( end );
 		// Unescape the parsed literal and handle escape sequences
-		final StringBuilder sb = new StringBuilder( text.length() - 2 );
-		for ( int i = 1; i < end; i++ ) {
+		final StringBuilder sb = new StringBuilder( text.length() - ( start + 2 ) );
+		for ( int i = start + 1; i < end; i++ ) {
 			char c = text.charAt( i );
 			switch ( c ) {
 				case '\'':
```

Comment:
The double unquoting should actually only happend when `!escape`, so I think it might be worth splitting the method into two. One that does just the double quote unescaping i.e. move logic to `unquoteStringLiteral`, and one that does just the escape sequence unescaping i.e. move logic to `unquoteJavaStringLiteral`

Reply:
Done!